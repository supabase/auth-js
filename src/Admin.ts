import { get } from './lib/fetch'
import { AdminConfig } from './Admin.types'

export default class Admin {
  url: string
  headers: AdminConfig['headers'] = {}

  /**
   * Creates a GoTrue instance for admin interactions.
   *
   * @param url  URL of the GoTrue instance.
   * @param headers  Custom headers.
   */
  constructor(url: string, options?: AdminConfig) {
    this.url = url
    if (options?.headers) {
      this.headers = { ...this.headers, ...options.headers }
    }
  }

  /**
   * Creates a new user account for your business or project.
   */
  async settings() {
    try {
      let data: any = await get(`${this.url}/settings`, { headers: this.headers })

      if (!data.external) {
        return { data: null, error: data.msg }
      }

      return { data, error: null }
    } catch (error) {
      return { data: null, error: error.toString() }
    }
  }
}
