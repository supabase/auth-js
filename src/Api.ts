import { get, post, put } from './lib/fetch'
import { Session, UserAttributes } from './lib/types'

export default class Api {
  url: string
  headers: {
    [key: string]: string
  }

  constructor({ url = '', headers = {} }: any) {
    this.url = url
    this.headers = headers
  }

  signUpWithEmail(email: string, password: string) {
    return post(`${this.url}/signup`, { email, password }, { headers: this.headers })
  }

  signInWithEmail(email: string, password: string) {
    return post(
      `${this.url}/token?grant_type=password`,
      { email, password },
      { headers: this.headers }
    )
  }

  resetPasswordForEmail(email: string) {
    return post(`${this.url}/forgotPassword`, { email }, { headers: this.headers })
  }

  async signOut(jwt: string) {
    let headers = { ...this.headers }
    headers['Authorization'] = `Bearer ${jwt}`
    console.log('headers', headers)
    let data = await post(`${this.url}/logout`, {}, { headers, noResolveJson: true })
    return data
  }

  getUser(jwt: string) {
    let headers = { ...this.headers }
    headers['Authorization'] = `Bearer ${jwt}`
    return get(`${this.url}/user`, { headers })
  }

  updateUser(jwt: string, attributes: UserAttributes) {
    let headers = { ...this.headers }
    headers['Authorization'] = `Bearer ${jwt}`
    return put(`${this.url}/user`, attributes, { headers })
  }

  refreshAccessToken(refreshToken: string) {
    return post(
      `${this.url}/token?grant_type=refresh_token`,
      { refresh_token: refreshToken },
      { headers: this.headers }
    )
  }
}
