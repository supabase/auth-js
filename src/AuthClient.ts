import GoTrueClient from './GoTrueClient'

const AuthClient = GoTrueClient

export default AuthClient
