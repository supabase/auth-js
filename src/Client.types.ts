export interface ClientConfig {
  headers?: {
    [key: string]: string
  }
}

/**
 * Available signup options. (More coming soon, like mobile..)
 */
export enum CredentialsType {
  email = 'EMAIL',
}

export interface Subscription {
  id: string
  callback: Function
  /** Removes the callback function */
  unsubscribe: Function
}

export interface UserCredentials {
  type: CredentialsType
  email?: string
  password?: string
}

export enum UserEvents {
  /** Triggered when a user signs up */
  SIGN_UP = 'SIGN_UP',
  /** Triggered when a user logs in */
  SIGN_IN = 'SIGN_IN',
  /** Triggered when a user logs out */
  SIGN_OUT = 'SIGN_OUT',
}
