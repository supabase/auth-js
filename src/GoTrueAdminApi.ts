import { Fetch, _generateLinkResponse, _request, _userResponse } from './lib/fetch'
import { resolveFetch } from './lib/helpers'
import {
  AdminUserAttributes,
  GenerateLinkParams,
  GenerateLinkResponse,
  User,
  UserResponse,
  GoTrueAdminMFAApi,
  AuthMFAAdminDeleteFactorParams,
  AuthMFAAdminDeleteFactorResponse,
  AuthMFAAdminListFactorsParams,
  AuthMFAAdminListFactorsResponse,
} from './lib/types'
import { AuthError, isAuthError } from './lib/errors'

export default class GoTrueAdminApi {
  /** Contains all MFA administration methods. */
  mfa: GoTrueAdminMFAApi

  protected url: string
  protected headers: {
    [key: string]: string
  }
  protected fetch: Fetch

  constructor({
    url = '',
    headers = {},
    fetch,
  }: {
    url: string
    headers?: {
      [key: string]: string
    }
    fetch?: Fetch
  }) {
    this.url = url
    this.headers = headers
    this.fetch = resolveFetch(fetch)
    this.mfa = {
      listFactors: this._listFactors.bind(this),
      deleteFactor: this._deleteFactor.bind(this),
    }
  }

  /**
   * Removes a logged-in session.
   * @param jwt A valid, logged-in JWT.
   */
  async signOut(jwt: string): Promise<{ data: null; error: AuthError | null }> {
    try {
      await _request(this.fetch, 'POST', `${this.url}/logout`, {
        headers: this.headers,
        jwt,
        noResolveJson: true,
      })
      return { data: null, error: null }
    } catch (error) {
      if (isAuthError(error)) {
        return { data: null, error }
      }

      throw error
    }
  }

  /**
   * Sends an invite link to an email address.
   * @param email The email address of the user.
   * @param options.redirectTo A URL or mobile deeplink to send the user to after they are confirmed.
   * @param options.data Optional user metadata
   */
  async inviteUserByEmail(
    email: string,
    options: {
      redirectTo?: string
      data?: object
    } = {}
  ): Promise<UserResponse> {
    try {
      return await _request(this.fetch, 'POST', `${this.url}/invite`, {
        body: { email, data: options.data },
        headers: this.headers,
        redirectTo: options.redirectTo,
        xform: _userResponse,
      })
    } catch (error) {
      if (isAuthError(error)) {
        return { data: { user: null }, error }
      }

      throw error
    }
  }

  /**
   * Generates email links and OTPs to be sent via a custom email provider.
   * @param email The user's email.
   * @param options.password User password. For signup only.
   * @param options.data Optional user metadata. For signup only.
   * @param options.redirectTo The redirect url which should be appended to the generated link
   */
  async generateLink(params: GenerateLinkParams): Promise<GenerateLinkResponse> {
    try {
      const { options, ...rest } = params
      const body: any = { ...rest, ...options }
      if ('newEmail' in rest) {
        // replace newEmail with new_email in request body
        body.new_email = rest?.newEmail
        delete body['newEmail']
      }
      return await _request(this.fetch, 'POST', `${this.url}/admin/generate_link`, {
        body: body,
        headers: this.headers,
        xform: _generateLinkResponse,
        redirectTo: options?.redirectTo,
      })
    } catch (error) {
      if (isAuthError(error)) {
        return {
          data: {
            properties: null,
            user: null,
          },
          error,
        }
      }
      throw error
    }
  }

  // User Admin API
  /**
   * Creates a new user.
   * This function should only be called on a server. Never expose your `service_role` key in the browser.
   */
  async createUser(attributes: AdminUserAttributes): Promise<UserResponse> {
    try {
      return await _request(this.fetch, 'POST', `${this.url}/admin/users`, {
        body: attributes,
        headers: this.headers,
        xform: _userResponse,
      })
    } catch (error) {
      if (isAuthError(error)) {
        return { data: { user: null }, error }
      }

      throw error
    }
  }

  /**
   * Get a list of users.
   *
   * This function should only be called on a server. Never expose your `service_role` key in the browser.
   */
  async listUsers(): Promise<
    { data: { users: User[] }; error: null } | { data: { users: [] }; error: AuthError }
  > {
    try {
      const { data, error } = await _request(this.fetch, 'GET', `${this.url}/admin/users`, {
        headers: this.headers,
      })
      if (error) throw error
      return { data: { ...data }, error: null }
    } catch (error) {
      if (isAuthError(error)) {
        return { data: { users: [] }, error }
      }
      throw error
    }
  }

  /**
   * Get user by id.
   *
   * @param uid The user's unique identifier
   *
   * This function should only be called on a server. Never expose your `service_role` key in the browser.
   */
  async getUserById(uid: string): Promise<UserResponse> {
    try {
      return await _request(this.fetch, 'GET', `${this.url}/admin/users/${uid}`, {
        headers: this.headers,
        xform: _userResponse,
      })
    } catch (error) {
      if (isAuthError(error)) {
        return { data: { user: null }, error }
      }

      throw error
    }
  }

  /**
   * Updates the user data.
   *
   * @param attributes The data you want to update.
   *
   * This function should only be called on a server. Never expose your `service_role` key in the browser.
   */
  async updateUserById(uid: string, attributes: AdminUserAttributes): Promise<UserResponse> {
    try {
      return await _request(this.fetch, 'PUT', `${this.url}/admin/users/${uid}`, {
        body: attributes,
        headers: this.headers,
        xform: _userResponse,
      })
    } catch (error) {
      if (isAuthError(error)) {
        return { data: { user: null }, error }
      }

      throw error
    }
  }

  /**
   * Delete a user. Requires a `service_role` key.
   *
   * @param id The user id you want to remove.
   *
   * This function should only be called on a server. Never expose your `service_role` key in the browser.
   */
  async deleteUser(id: string): Promise<UserResponse> {
    try {
      return await _request(this.fetch, 'DELETE', `${this.url}/admin/users/${id}`, {
        headers: this.headers,
        xform: _userResponse,
      })
    } catch (error) {
      if (isAuthError(error)) {
        return { data: { user: null }, error }
      }

      throw error
    }
  }

  private async _listFactors(
    params: AuthMFAAdminListFactorsParams
  ): Promise<AuthMFAAdminListFactorsResponse> {
    try {
      const data = await _request(
        this.fetch,
        'GET',
        `${this.url}/admin/users/${params.userId}/factors`,
        {
          headers: this.headers,
        }
      )
      return { data, error: null }
    } catch (error) {
      if (isAuthError(error)) {
        return { data: null, error }
      }

      throw error
    }
  }

  private async _deleteFactor(
    params: AuthMFAAdminDeleteFactorParams
  ): Promise<AuthMFAAdminDeleteFactorResponse> {
    try {
      const data = await _request(
        this.fetch,
        'DELETE',
        `${this.url}/admin/users/${params.userId}/factors/${params.id}`,
        {
          headers: this.headers,
        }
      )

      return { data, error: null }
    } catch (error) {
      if (isAuthError(error)) {
        return { data: null, error }
      }

      throw error
    }
  }
}
