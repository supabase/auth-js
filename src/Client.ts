import Api from './Api'
import { isBrowser } from './lib/helpers'
import { STORAGE_KEY } from './lib/constants'
import { GOTRUE_URL, DEFAULT_HEADERS } from './lib/constants'
import { Session, User, UserAttributes } from './lib/types'

export default class Client {
  api: Api
  currentUser: User | null
  currentSession?: Session | null
  autoRefreshToken: boolean
  persistSession: boolean

  constructor({
    url = GOTRUE_URL,
    autoRefreshToken = true,
    persistSession = true,
    headers = DEFAULT_HEADERS,
  }: any) {
    this.currentUser = null
    this.currentSession = null
    this.autoRefreshToken = autoRefreshToken
    this.persistSession = persistSession
    this.api = new Api({ url, headers })
    this._recoverSession()
  }

  async signUp({ email, password }: { email: string; password: string }) {
    try {
      this._removeSession()

      let data: any = await this.api.signUpWithEmail(email, password)

      if (data?.user?.confirmed_at) this._saveSession(data)

      return { data, error: null }
    } catch (error) {
      return { data: null, error: error.toString() }
    }
  }

  async signIn({ email, password }: { email: string; password: string }) {
    try {
      this._removeSession()

      let data: any = await this.api.signInWithEmail(email, password)

      if (data?.user?.confirmed_at) this._saveSession(data)

      return { data, error: null }
    } catch (error) {
      return { data: null, error: error.toString() }
    }
  }

  async user() {
    try {
      if (!this.currentSession?.access_token) throw new Error('Not logged in.')
      let data: any = await this.api.getUser(this.currentSession.access_token)
      this.currentUser = data
      return { data: this.currentUser, error: null }
    } catch (error) {
      return { data: null, error: error.toString() }
    }
  }

  async update(attributes: UserAttributes) {
    try {
      if (!this.currentSession?.access_token) throw new Error('Not logged in.')
      let data: any = await this.api.updateUser(this.currentSession.access_token, attributes)
      this.currentUser = data
      return { data: this.currentUser, error: null }
    } catch (error) {
      return { data: null, error: error.toString() }
    }
  }

  async signOut() {
    try {
      if (this.currentSession) {
        await this.api.signOut(this.currentSession.access_token)
      }
      this._removeSession()
      return true
    } catch (error) {
      return { data: null, error: error.toString() }
    }
  }

  _saveSession(session: Session) {
    this.currentSession = session
    this.currentUser = session['user']
    let tokenExpirySeconds = session['expires_in']

    if (this.autoRefreshToken && tokenExpirySeconds) {
      setTimeout(this._callRefreshToken, (tokenExpirySeconds - 60) * 1000)
    }

    if (this.persistSession) {
      this._persistSession(this.currentSession, this.currentUser, tokenExpirySeconds)
    }
  }

  _persistSession(currentSession: Session, currentUser: User, secondsToExpiry: number) {
    const timeNow = Math.round(Date.now() / 1000)
    const expiresAt = timeNow + secondsToExpiry
    const data = { currentSession, currentUser, expiresAt }
    isBrowser() && localStorage.setItem(STORAGE_KEY, JSON.stringify(data))
  }

  _removeSession() {
    this.currentSession = null
    this.currentUser = null
    isBrowser() && localStorage.removeItem(STORAGE_KEY)
  }

  _recoverSession() {
    const json = isBrowser() && localStorage.getItem(STORAGE_KEY)
    if (json) {
      try {
        const data = JSON.parse(json)
        const { currentSession, currentUser, expiresAt } = data

        const timeNow = Math.round(Date.now() / 1000)
        if (expiresAt < timeNow) {
          console.log('Saved session has expired.')
          this._removeSession()
        } else {
          this.currentSession = currentSession
          this.currentUser = currentUser
          // schedule a refresh 60 seconds before token due to expire
          setTimeout(this._callRefreshToken, (expiresAt - timeNow - 60) * 1000)
        }
      } catch (err) {
        console.error(err)
        return null
      }
    }
    return null
  }

  async _callRefreshToken() {
    try {
      if (this.currentSession?.refresh_token) {
        let data: any = await this.api.refreshAccessToken(this.currentSession?.refresh_token)

        if (data?.access_token) {
          this.currentSession.access_token = data.body['access_token']
          this.currentSession.refresh_token = data.body['refresh_token']
          let tokenExpirySeconds = data.body['expires_in']

          if (this.autoRefreshToken && tokenExpirySeconds) {
            setTimeout(this._callRefreshToken, (tokenExpirySeconds - 60) * 1000)
          }

          if (this.persistSession && this.currentUser) {
            this._persistSession(this.currentSession, this.currentUser, tokenExpirySeconds)
          }
        }
        return { data, error: null }
      }
    } catch (error) {
      return { data: null, error: error.toString() }
    }
  }
}
