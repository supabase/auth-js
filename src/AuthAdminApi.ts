import GoTrueAdminApi from './GoTrueAdminApi'

const AuthAdminApi = GoTrueAdminApi

export default AuthAdminApi
