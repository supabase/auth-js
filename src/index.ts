import GoTrueAdminApi from './GoTrueAdminApi'
import GoTrueClient from './GoTrueClient'
export { GoTrueAdminApi, GoTrueClient }
export * from './lib/types'
export * from './lib/errors'
export {
  navigatorLock,
  NavigatorLockAcquireTimeoutError,
  internals as lockInternals,
} from './lib/locks'
