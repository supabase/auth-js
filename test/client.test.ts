import { Client } from '../src/index'

const GOTRUE_URL = 'http://localhost:9999'

const auth = new Client({
  url: GOTRUE_URL,
  autoRefreshToken: false,
  persistSession: true,
})

const email = 'fake@email.com'
const password = 'secret'

test('signUp()', async () => {
  let { error, data } = await auth.signUp({
    email,
    password,
  })
  expect(error).toBeNull()
  expect(data).toMatchSnapshot({
    access_token: expect.any(String),
    expires_in: expect.any(Number),
    user: {
      id: expect.any(String),
      aud: expect.any(String),
      updated_at: expect.any(String),
      app_metadata: {
        provider: 'email',
      },
    },
  })
})

test('signIn()', async () => {
  let { error, data } = await auth.signIn({
    email,
    password,
  })
  expect(error).toBeNull()
  expect(data).toMatchSnapshot({
    access_token: expect.any(String),
    expires_in: expect.any(Number),
    user: {
      id: expect.any(String),
      aud: expect.any(String),
      updated_at: expect.any(String),
      app_metadata: {
        provider: 'email',
      },
    },
  })
})

test('Get user', async () => {
  let { error, data } = await auth.user()
  expect(error).toBeNull()
  expect(data).toMatchSnapshot({
    id: expect.any(String),
    aud: expect.any(String),
    updated_at: expect.any(String),
    app_metadata: {
      provider: 'email',
    },
  })
})

test('Update user', async () => {
  let { error, data } = await auth.update({ data: { hello: 'world' } })
  expect(error).toBeNull()
  expect(data).toMatchSnapshot({
    id: expect.any(String),
    aud: expect.any(String),
    updated_at: expect.any(String),
    user_metadata: {
      hello: 'world',
    },
  })
})

test('Get user after updating', async () => {
  let { error, data } = await auth.user()
  expect(error).toBeNull()
  expect(data).toMatchSnapshot({
    id: expect.any(String),
    aud: expect.any(String),
    updated_at: expect.any(String),
    user_metadata: {
      hello: 'world',
    },
  })
})

test('signOut', async () => {
  let res = await auth.signOut()
  expect(res).toMatchSnapshot()
})

test('Get user after logging out', async () => {
  let res = await auth.user()
  expect(res).toMatchSnapshot()
})
