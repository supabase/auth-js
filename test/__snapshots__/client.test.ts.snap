// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Get user 1`] = `
Object {
  "app_metadata": Object {
    "provider": "email",
  },
  "aud": Any<String>,
  "confirmed_at": "2020-09-22T11:11:48.811209Z",
  "created_at": "2020-09-22T11:11:48.807848Z",
  "email": "fake@email.com",
  "id": Any<String>,
  "role": "",
  "updated_at": Any<String>,
  "user_metadata": Object {},
}
`;

exports[`Get user after logging out 1`] = `
Object {
  "data": null,
  "error": "Error: Not logged in.",
}
`;

exports[`Get user after updating 1`] = `
Object {
  "app_metadata": Object {
    "provider": "email",
  },
  "aud": Any<String>,
  "confirmed_at": "2020-09-22T11:11:48.811209Z",
  "created_at": "2020-09-22T11:11:48.807848Z",
  "email": "fake@email.com",
  "id": Any<String>,
  "role": "",
  "updated_at": Any<String>,
  "user_metadata": Object {
    "hello": "world",
  },
}
`;

exports[`Update user 1`] = `
Object {
  "app_metadata": Object {
    "provider": "email",
  },
  "aud": Any<String>,
  "confirmed_at": "2020-09-22T11:11:48.811209Z",
  "created_at": "2020-09-22T11:11:48.807848Z",
  "email": "fake@email.com",
  "id": Any<String>,
  "role": "",
  "updated_at": Any<String>,
  "user_metadata": Object {
    "hello": "world",
  },
}
`;

exports[`signIn() 1`] = `
Object {
  "access_token": Any<String>,
  "expires_in": Any<Number>,
  "refresh_token": "2XeSL-nTVOZCsG6PKfKhVg",
  "token_type": "bearer",
  "user": Object {
    "app_metadata": Object {
      "provider": "email",
    },
    "aud": Any<String>,
    "confirmed_at": "2020-09-22T11:11:48.811209Z",
    "created_at": "2020-09-22T11:11:48.807848Z",
    "email": "fake@email.com",
    "id": Any<String>,
    "last_sign_in_at": "2020-09-22T11:11:48.9022814Z",
    "role": "",
    "updated_at": Any<String>,
    "user_metadata": Object {},
  },
}
`;

exports[`signOut 1`] = `true`;

exports[`signUp() 1`] = `
Object {
  "access_token": Any<String>,
  "expires_in": Any<Number>,
  "refresh_token": "GSiVzIcREbjs9mGn6knGeA",
  "token_type": "bearer",
  "user": Object {
    "app_metadata": Object {
      "provider": "email",
    },
    "aud": Any<String>,
    "confirmed_at": "2020-09-22T11:11:48.8112086Z",
    "created_at": "2020-09-22T11:11:48.807848Z",
    "email": "fake@email.com",
    "id": Any<String>,
    "last_sign_in_at": "2020-09-22T11:11:48.8142114Z",
    "role": "",
    "updated_at": Any<String>,
    "user_metadata": null,
  },
}
`;
